### PR TITLE
mm: Fix memory corruption issues in realloc

### PR DESCRIPTION
**Pull request by yamt**
Also, add a bunch of assertions.

### CODE REVIEW

**Comment by yamt** [comment]
hm, it seems the real issue was not here. MM_MIN_CHUNK was broken for clang. i'm closing this for now.

File: mm/mm_heap/mm_realloc.c [hunk 1]
```diff
@@ -213,6 +215,28 @@ FAR void *mm_realloc(FAR struct mm_heap_s *heap, FAR void *oldmem,
...
         }
 
+      /* Note that a free node can't be smaller than SIZEOF_MM_FREENODE.
+       * We have to either take the whole node, or leave at least
+       * SIZEOF_MM_FREENODE bytes.
+       * When we can't leave SIZEOF_MM_FREENODE bytes, just take the whole
+       * node. It might end up with returning a larger chunk than the
+       * caller requested. It should be ok.
+       */
+      DEBUGASSERT(prevsize >= takeprev);
```
Comment (patacongo) [inline mm/mm_heap/mm_realloc.c]: Coding standard requires that all comments be followed by a blank line.

File: mm/mm_heap/mm_realloc.c [hunk 2]
```diff
@@ -213,6 +215,28 @@ FAR void *mm_realloc(FAR struct mm_heap_s *heap, FAR void *oldmem,
...
+      DEBUGASSERT(nextsize >= takenext);
+      if (prevsize - takeprev < SIZEOF_MM_FREENODE)
+        {
+          takeprev = prevsize;
+        }
+      if (nextsize - takenext < SIZEOF_MM_FREENODE)
```
Comment (patacongo) [inline mm/mm_heap/mm_realloc.c]: Coding standard requires that all terminating right braces be followed by a blank line.

File: mm/mm_heap/mm_realloc.c [hunk 3]
```diff
@@ -213,6 +215,28 @@ FAR void *mm_realloc(FAR struct mm_heap_s *heap, FAR void *oldmem,
...
+        {
+          takenext = nextsize;
+        }
+      DEBUGASSERT(prevsize == takeprev ||
```
Comment (patacongo) [inline mm/mm_heap/mm_realloc.c]: Needs a blank line

**Pull request closed without merging**